agent/xds/server.go

```diff
@@ -0,0 +1,418 @@
+	return errors.New("not implemented")
+}
+
+func deniedResponse(reason string) (*authz.CheckResponse, error) {
+	return &authz.CheckResponse{
+		Status: &rpc.Status{
+			Code:    int32(rpc.PERMISSION_DENIED),
+			Message: "Denied: " + reason,
+		},
+	}, nil
+}
+
+// Check implementents authz.AuthorizationServer.
+func (s *Server) Check(ctx context.Context, r *authz.CheckRequest) (*authz.CheckResponse, error) {
+	// Sanity checks
+	if r.Attributes == nil || r.Attributes.Source == nil || r.Attributes.Destination == nil {
+		return nil, status.Error(codes.InvalidArgument, "source and destination attributes are required")
+	}
+	if r.Attributes.Source.Principal == "" || r.Attributes.Destination.Principal == "" {
+		return nil, status.Error(codes.InvalidArgument, "source and destination Principal are required")
+	}
+
+	// Parse destination to know the target service
+	dest, err := connect.ParseCertURIFromString(r.Attributes.Destination.Principal)
+	if err != nil {
+		// Treat this as an auth error since Envoy has sent something it considers
+		// valid, it's just not an identity we trust.
+		return deniedResponse("Destination Principal is not a valid Connect identitiy")
+	}
+
+	destID, ok := dest.(*connect.SpiffeIDService)
+	if !ok {
+		return deniedResponse("Destination Principal is not a valid Service identitiy")
+	}
+
+	// For now we don't validate the trust domain of the _destination_ at all -
+	// the HTTP Authorize endpoint just accepts a target _service_ and it's
+	// implicit that the request is for the correct cluster. We might want to
+	// reconsider this later but plumbing in additional machinery to check the
+	// clusterID here is not really necessary for now unless Envoys are badly
+	// configured. Our threat model _requires_ correctly configured and well
+	// behaved proxies given that they have ACLs to fetch certs and so can do
+	// whatever they want including not authorizing traffic at all or routing it
+	// do a different service than they authed against.
+
+	// Create an authz request
+	req := &structs.ConnectAuthorizeRequest{
+		Target:        destID.Service,
+		ClientCertURI: r.Attributes.Source.Principal,
+		// TODO(banks): need Envoy to support sending cert serial/hash to enforce
+		// revocation later.
+	}
+	token := tokenFromContext(ctx)
+	authed, reason, _, err := s.authz.ConnectAuthorize(token, req)
+	if err != nil {
+		if err == acl.ErrPermissionDenied {
+			return nil, status.Error(codes.PermissionDenied, err.Error())
+		}
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+	if !authed {
+		return deniedResponse(reason)
+	}
+
+	return &authz.CheckResponse{
+		Status: &rpc.Status{
+			Code:    int32(rpc.OK),
+			Message: "ALLOWED: " + reason,
+		},
+	}, nil
+}
+
+// GRPCServer returns a server instance that can handle XDS and ext_authz
+// requests.
+func (s *Server) GRPCServer(certFile, keyFile string) (*grpc.Server, error) {
+	opts := []grpc.ServerOption{
+		grpc.MaxConcurrentStreams(2048),
```

Comment:
Why 2048?

Reply:
Seems like enough :)

The docs are super unclear but it was received wisdom for a while at least that gRPC limits to 100 streams by default which is probably enough for us but just in case you have more than 100 envoys talking to one agent I bumped it up. For example: https://github.com/cockroachdb/cockroach/issues/14769 

The current code doesn't _seem_ to default to 100 anywhere from a quick look but this was just a reasonable number in general.